jdaviz/core/template_mixin.py

```diff
@@ -436,3 +438,165 @@ def __init__(self, *args, **kwargs):
                                            'spatial_subset_selected_has_subregions',
                                            default_text='No Subset',
                                            allowed_type='spatial')
+
+
+class ViewerSelect(BasePluginComponent):
+    """
+    Traitlets (in the object, custom traitlets in the plugin):
+
+    * ``items`` (list of dicts with keys: id, reference, ref_or_id)
+    * ``selected`` (string)
+
+    Properties (in the object only):
+
+    * ``ids`` (list of ids corresponding to ``items``)
+    * ``references`` (list of references corresponding to ``items``)
+    * ``ref_or_ids`` (list of references falling back on ids corresponding to ``items``.  These
+        are the values seen in the dropdown, although setting either id or reference to the traitlet
+        will still process correctly)
+    * ``selected_item`` (dict of the currently selected entry in ``items``)
+    * ``selected_id`` (string corresponding to the id of ``selected_item``)
+    * ``selected_obj`` (viewer item corresponding to ``selected``)
+
+
+    To use in a plugin:
+
+    * create traitlets with default values
+    * register with all the automatic logic in the plugin's init by passing the string names
+      of the respective traitlets.
+    * use component in plugin template (see below)
+    * refer to properties above based on the interally stored reference to the
+      instantiated object of this component
+
+    Example template (label and hint are optional)::
+
+      <plugin-viewer-select
+        :items="viewer_items"
+        :selected.sync="viewer_selected"
+        label="Viewer"
+        hint="Select viewer."
+      />
+
+    """
+    def __init__(self, plugin, items, selected):
+        super().__init__(plugin,
+                         items=items,
+                         selected=selected)
+
+        self.hub.subscribe(self, ViewerAddedMessage,
+                           handler=lambda _: self._on_viewers_changed())
+        self.hub.subscribe(self, ViewerRemovedMessage,
+                           handler=lambda _: self._on_viewers_changed())
+        self.add_observe(selected, self._selected_changed)
+
+        # initialize viewer_items from original viewers
+        self._on_viewers_changed()
+
+    @property
+    def ids(self):
+        return [item['id'] for item in self.items]
+
+    @property
+    def references(self):
+        return [item['reference'] for item in self.items]
+
+    @property
+    def ref_or_ids(self):
+        return [item['ref_or_id'] for item in self.items]
+
+    @property
+    def selected_item(self):
+        for item in self.items:
+            if item['ref_or_id'] == self.selected:
+                return item
+        # try again but this time allow match to id alone.  Note that _selected_changed
```

Comment:
Why is this needed? Looks like the tests don't go here according to codecov.

Reply:
This is needed in order to support setting the selected item based on the id.  I think its worth keeping, so will add tests to cover it.